test/unit/query/builder.js

```diff
@@ -848,6 +848,66 @@ describe('QueryBuilder', () => {
     });
   });
 
+  it('basic wheres should not accept array or object as a value #1227', () => {
+    testquery(qb().select('*').from('users').where('id', '=', 1), {
+      mysql: 'select * from `users` where `id` = 1',
+      pg: 'select * from "users" where "id" = 1',
+      'pg-redshift': 'select * from "users" where "id" = 1',
+      mssql: 'select * from [users] where [id] = 1',
```

Comment:
Instead of "= 1", it should be "= \`1\`" because for mysql, if `= 0` is used, it may match other values. I'm not sure if it should be this way for other databases though, can someone versed in the other databases comment on this?

Reply:
I tested this on db-fiddle, postgre does not allow comparisons between types with `=` operator and expectedly returns an error.

Reply:
Yeah, this is scary but it is rather a MySQL issue than a library concern.  Or at least this is a separate case with the type coercion in MySQL while this pr handles the invalid usage of the where clause with objects as plain values.

Reply:
> Yeah, this is scary but it is rather a MySQL issue than a library concern.

I disagree that this is not a library concern. The purpose of knex is to abstract away database-specific quirks to allow safe query-building. Since mysql has the quirk of "= 0" matching all values, knex is responsible for escaping it to "= \`0\`" to represent the developer's intent.

> Or at least this is a separate case with the type coercion in MySQL while this pr handles the invalid usage of the where clause with objects as plain values.

Should adding backticks around values for mysql be added in a different pr then? Two prs addressing the fundamental issue of knex not expecting non-strings through assertions and escaping seems unnecessarily complex, error-prone, and time-consuming compared to calling `.toString()` on inputs.

Reply:
> Since mysql has the quirk of "= 0" matching all values, knex is responsible for escaping it to "= `0`" to represent the developer's intent.

@r2dev2 there are some issues with this.
1 - backticks for values is invalid syntax in MySQL, backticks should only be used for identifiers.
2 - when we consider quotes instead of backticks, evne this is a bit contentious, I thought it would break with integers, but MySQL is full of surprises. `= 0` is perfectly valid in case of integers, and for reasonable types, probably should be used. 
Here are the tests related tests: https://www.db-fiddle.com/f/vytxD1kFa3xBiPf7TBADxq/4

![image](https://user-images.githubusercontent.com/5603529/210073760-41940404-aee3-42a9-b534-358098124594.png)

Reply:
> 1 - backticks for values is invalid syntax in MySQL, backticks should only be used for identifiers

Ah, it appears I got a bit confused with previous experiments. It is the quote that should be used for values, not backtick. I added a quote escaping test to your fiddle (so it does `= '0'` instead of `= 0` for secret column) and that works properly.

Fiddle: https://www.db-fiddle.com/f/vytxD1kFa3xBiPf7TBADxq/5

Reply:
The change is not simple toString. In raw queries you need to convert bindings to string as well, while the collection of bindings could be either array or object (named bindings)

Reply:
> The change is not simple toString. In raw queries you need to convert bindings to string as well, while the collection of bindings could be either array or object (named bindings)

Yea, I tried calling `.toString()` at various places in the codebase and it seems like calling `.toString()` on all the bindings is not a solution as a binding could map to many different situations such as `limit = ?`

Perhaps only calling `.toString()` for `where =` clauses and other operators where the value can be quoted is a solution? One dirty way I did this was by adding

```js
whereBasic(statement) {
  if (['number', 'boolean', 'array', 'object'].includes(typeof statement.value)) {
    statement.value = statement.value.toString();
  }
  return super.whereBasic(statement);
}
```

to `lib/dialects/mysql/query/mysql-querycompiler.js`

However, I'm sure there is a better way of doing this which someone more familiar with the knex codebase than me can think of.